scripts/kconfig/kconfig.py

```diff
@@ -1,22 +1,33 @@
 #!/usr/bin/env python3
 # Modified from: https://github.com/ulfalizer/Kconfiglib/blob/master/examples/merge_config.py
 from kconfiglib import Kconfig, Symbol, BOOL, STRING, TRISTATE, TRI_TO_STR
+import io
 import sys
 
 if len(sys.argv) < 5:
     print('usage: {} Kconfig dotconfig autoconf conf1 [conf2 ...]'
           .format(sys.argv[0]))
     sys.exit(1)
 
+
+# Capture warnings from Kconfiglib, which always go to stderr, into a string.
+# This allows us to turn some of them into errors later.
+#
+# This is a bit hacky, but I'm not sure whether it's worth adding a less
+# dead-simple warning scheme to Kconfiglib for this 1% case.
```

Comment:
I'm concerned about how well this will work into the future. What do you think about
having an API-equivalent of what gcc does?

From what I have seen of how kconfiglib treats various weird user-input. I suspect that Zephyr will
in general want to be more strict about what is legal and in many cases prefer to error-out
where Linux would prefer to warn (presumably for compatibility reasons).

> -w
> Inhibit all warning messages.
> 
> -Werror
> Make all warnings into errors.
> 
> -Werror=
> Make the specified warning into an error. The specifier for a warning is appended; for example -Werror=switch turns the warnings controlled by -Wswitch into errors. This switch takes a negative form, to be used to negate -Werror for specific warnings; for example -Wno-error=switch makes -Wswitch warnings not be errors, even when -Werror is in effect.
> 
> The warning message for each controllable warning includes the option that controls the warning. That option can then be used with -Werror= and -Wno-error= as described above. (Printing of the option in the warning message can be disabled using the -fno-diagnostics-show-option flag.)
> 
> Note that specifying -Werror=foo automatically implies -Wfoo. However, -Wno-error=foo does not imply anything.
> 
> -Wfatal-errors
> This option causes the compiler to abort compilation on the first error occurred rather than trying to keep going and printing further error messages.

Reply:
Yeah, some slightly fancier scheme might be nice if more cases like this come up.

I've been thinking of optionally making the warnings available as a list instead of hardcoding them to go to `stderr` (though that would stll be the default). That would have pretty good bang-for-the-complexity I think.

I prefer to keep error policy out of Kconfiglib itself. Unless you go for some simple `-Werror` thing (which would be simple to implement with the list, or by capturing `stderr`), you'd need to label up every single warning in the library and document them somehow. That would get painful I think.

If you were to label up the warnings, I think what I would go for would be to return the labels in the list, leaving the handling up to the application in that way.

Implementation-wise, it would make the code more complex if you suddenly had to deal with errors in a bunch of places that previously generated warnings, because I wouldn't want the state to be potentially messed-up after the exception.

I agree that that `stderr` override isn't the prettiest thing in the world though. If a lot of other stuff like this comes up, I'd definitely want *something*. :)

Reply:
"Implementation-wise, it would make the code more complex if you suddenly had to deal with errors in a bunch of places that previously generated warnings, because I wouldn't want the state to be potentially messed-up after the exception."

perhaps stderr like gcc does is okay then, the API only takes input about how to react to errors and warnings, it doesn't change it's control flow unless the error requires it to do so. This would meet Zephyr's requirements at least.

Reply:
If you wanted to make the warning detection guaranteed reliable in case the text changes, you could tag the warnings in some way, either on `stderr` or just in the warning-fetching API. I have a feeling that might be overkilling it for Kconfiglib though.

Reply:
Hmm... you might have been agreeing with me. Sorry for the rant in that case. :)

Reply:
End of the work-day here, so I'm having trouble understanding. :)

But I was trying to help you out with the "exceptions break control flow" problem, by suggesting that it could work like this instead:

When invoking the API the caller decides which warnings are disabled, enabled, errored.

kconfiglib will either suppress, indicate a warning on STDOUT/STDERR, or indicate an error on STDERR based on the warning settings. Kconfiglib will also eventually (or immediately if necessary) return an error code that kconfig.py can use to appropriately fail the build if an error occured.

This design would mean kconfig.py would behave just like gcc does, and gcc has proven to have a usable warning system.